README.TXT
#                _____              ___ ____     _ _
#               |  ___| __ ___  ___|_ _|  _ \ __| | |__
#               | |_ | '__/ _ \/ _ \| || |_) / _` | '_ \
#               |  _|| | |  __/  __/| ||  __/ (_| | |_) |
#               |_|  |_|  \___|\___|___|_|   \__,_|_.__/
#
#	README.TXT-$Name:  $-$Revision: 1.9 $ $Date: 2002/05/07 23:30:55 $ <$Author: bapril $@freeipdb.org>
######################################################################

1  What is FreeIPdb?
2  Features.
3  How does FreeIPdb Work?
4  What are the requirements to run FreeIPdb?
5  How do I get Help with FreeIPdb?
6  How do I drive FreeIPdb?

1		What is FreeIPdb.

2		Features

       Regions: A Region is used to divide or combine portions of 
       	IP space into a independent sections. This gives the 
	operator the ability to either keep IP space separate 
	(i.e. RIPE,APNIC,ARIN), Allocate a subnet to a project 
	or customer that would then be allowed to assign subnets 
	from that subnet in any manner they choose. (i.e. operator 
	allocates a /24 to the new "shiny-widget" project. The 
	manager of the "shiny-widget" project chooses to assign 
	a /25 to the "widgets" and a /26 to the "shiny things" in 
	L.A. and a /27 to the "shiny things" in New York. The 
	manager can see that he has a /27 left and the operator 
	knows exactly how the project manager used the IP space that
	he was assigned.) Two different regions are allowed to contain 
	IP space from the same block (i.e. more than one region of 
	RFC-1918 space).  
       Reclaim: When a block (or a parent of it) is in reclaim mode 
	that block will NOT be re-allocated or assigned from the 
	database in the event that it is reclaimed. This feature 
	is useful if the operator has a need to return a block of 
	IP space to a RIR or any situation where it is not desirable 
	for a block to be allocated or assigned. 
       Priorities: As supernets are added to the database for the 
	first time the operator may assign a priority. A priority 
	is a value that can be used by the database to assign/allocate 
	space from one supernet before another. 
       IPv6: FreeIPdb works seamlessly with IPv6. The only restriction 
	is that IPv6 space be added in a region that contains no IPv4 
	Space. Other than that all features work the same way. 
       Hold Time: The operator may set a global hold time (this will 
       	soon be per region). Every time a block is reclaimed a 
	datestamp is set in the block (current time + holdime) when 
	that datestamp is less then the current time the database 
	will once again allow that block to be allocated/assigned or 
	rolled up into it's parent. 
       Set a specific block: In the event the operator needs to 
	allocate/assign an exact block (i.e. not whatever the database 
	returns) they can use the "Set a specific block" feature. The 
	database will automaticly create all the parents and children 
	needed to locate this block in the database. 

3               How does FreeIPdb work?

       FreeIPdb is based on a binary tree database. Every block can 
       be the parent of 2 children (A /24 is the parent of 2 /25's). 
       FreeIPdb is written in Perl to be universally portable. 
       All IP addresses (v4 and v6) are stored in the database as integers 
       The allocation and reclaim routines are fully recursive. 
       If both children of a parent are not in use both children 
       will be pruned from the tree and the parent
       etc... 
       The database uses an algorithm of best fit. If you do nothing 
       but assignment/allocation(s) you will never have more than 
       one unused block of each size in that given region at any time. 



4               What are the requirements to run FreeIPdb?
       A Perl interpreter. (tested on 5.005_03 built for i386-freebsd). 
       PostgreSQL 7.0 with Perl interface installed. (also tested on 7.1.2). 
       Math::BigInt-1.42 We're dealing with some mind-boggling-ly big numbers. 
       Net::IP-1.0 
       A Webserver capable of running Perl cgi (If you want the web front-end.) 
       Net::Telnet (for the route announcment check).


5               How do I get Help with FreeIPdb

       Send an e-mail to the addresses below with 'subscribe' in the subject 
       (you should know how to do this by now;-) 
       Announce - announce-request@FreeIPdb.org - 
       		New releases and important updates. 
       Users - users-request@FreeIPdb.org - 
       		User questions and support. 
       Hackers - hackers-request@FreeIPdb.org - 
       		Development and customization.

6  		How do I drive FreeIPdb?
	
	I'm glad you asked that! Lets start with the admin.cgi page.
	It should be something like cgi-bin/freeipdb-?_?_?/admin.cgi.
	This is where you will add new Registrars, Regions, and Allocations.
	The first thing we need is a Registrar(RA). We'll use ARIN for our
	example. You could use RIPE,APNIC,My_RAR,RFC1918 whatever you want.
	A RA is used to keep track of which space comes from which Authority.
	This way when it comes time to go get more we can do some easy reports
	and poof you're done. So find the section titled "Add a RA", fill in
	the field and click the [Add RA] Button. Provided you are all set
	up correctly you should see the same page with "RA <name> Added" at
	the top. The "RA:" Pull-down in the "Add A Region" section should 
	now list your RA.

	Our next step to getting FreeIPdb to do something useful will be to
	add a region. 

	<Tangent>
	What the heck is a region?
	
	A region is a "Pool" of IP space within an RA that may be either
	IPv4 or IPv6(not both). All IP space requests to FreeIPdb require
	a region and will return space from that region. You will likely 
	want to think a little about how you want to define your regions. 
	Some sample region schemes:

		US/MEXICO/CANADA
		North Am./South Am./Asia/Europe/Aust.
		MA/NH/ME/VT/CT/RI
		East/Central/West
		Servers/Desktops/WAN/SAN

	I'll let you go on from there. One way to think of a region is like
	a disk partition. You give 30GB to var and it's used for one type of
	operation another 20GB to /usr and that's for different operations.
	/usr can fill up and not have any adverse effect on /var.
	</Tangent>

	So now find the "Add a Region" section of the admin page. Choose an
	RA from the pull-down. Give your region a name. If you will be putting
	IPv6 IP space in this region check the box. and hit [Add Region]. 
	You should come back to the same page only the region you just added 
	will be in the "Region:" pull-downs.

	Now for the last step before we can get some gratification :-)
	Add an Allocation. Find the "Add Allocation" section. Enter the 
	name of the block so 10.10.10.0 not 10.10.10.0/24. Now the netmask
	so 24 for a class C. Choose a region to add this block to.

	<Tangent>
	What the heck is a priority?

	I use priorities to keep my older allocation at as close to 100%
	use as I can. The lower the number the more likely the block
	is to be used. If the database can find a larger block with a
	lower priority than a block of the correct size it will split the
	block with the better priority. The tool will do anything possible 
	to use a block with a lower priority.
		
	If priority is left null the block will be the last block to be used
	provided there are block with real priority values in that region.
	More than one block with no priority will get used in a quasi-random
	fashion.
	</Tangent>

	Enter a priority. Click [Add]. It should be in there. Now we can
	have some real fun.

	Lets move to the index.cgi. This page is where the day to day
	user functions will take place. I want to see that the Allocation
	we just added made it to the database. Choose "List free blocks"
	from the action pull-down and click [Submit Request].
	You should see a list of all the free blocks in your database. 
	(If you are following along there should be one entry.)
	
	Lets get a block. Go back to the index.cgi page. Set Action to 
	"Request Assignment". Choose the size of the block you want say a 
	/28 or 16 IP addresses. Select a region. You can use the next two 
	fields for some customer or internal tracking information. now click
	[Submit Request]. You should get something like this:

Your New BlockID is : 327 
10.0.23.208/28 from NorthAmerica To: 12333 IPv4 

[Confirm Allocation] [Cancel Allocation]

	If you click the Confirm button the block will be saved as in-use
	if you clock cancel it will be reclaimed and become available again.
	(Note: the browser back button will leave this block in an in-use state)

	Lets do one more thing before we go. Reclaim a block. To do this we 
	need to find the block. we know the the block is a /28 and the region
	is <what ever you choose> so fill them in. Choose "Search for block in 
	database" in the action field and click [Submit Request]. You should
	see your block and maybe others in the list.  Here you can either edit
	or reclaim the block.

	<Tangent>
	A word about reclaims:

	Reclaiming can be fun. It can also cause lots of trouble. Customer
	who think they are done with a block can and do turn-around 10 days
	later and say the need it back. This does not make the Database guy's
	job any easier. We have included the "10 day holding period". When a 
	block is reclaimed it becomes ready to use that instant however there
	is a flag set on that entry so that it will not be reused until the
	10-days have elapsed. (You can use the admin.cgi page to manually 
	assign that block if the customer wants it back.)
	BTW the holdtime is an option in the config.pm file
	</Tangent>

	Thats is for now. 
	
More to come. (wanna contribute? ;-)


CHANGES.TXT
#                _____              ___ ____     _ _
#               |  ___| __ ___  ___|_ _|  _ \ __| | |__
#               | |_ | '__/ _ \/ _ \| || |_) / _` | '_ \
#               |  _|| | |  __/  __/| ||  __/ (_| | |_) |
#               |_|  |_|  \___|\___|___|_|   \__,_|_.__/
#
#	CHANGES.TXT-$Name:  $-$Revision: 1.9 $ $Date: 2002/04/03 20:58:41 $ <$Author: bapril $@freeipdb.org>
######################################################################


0_1_1
	-Fixed a pile of Data_Types.txt issues.
		Dominic J. Eidson
		Jason Slagle
	
0_1_2
	-Added password to the config.pm file
		Jason Slagle
	-Started INSTALL.TXT,README.TXT and CHANGES.TXT
	-rev_gen_tool.pl now created reverse files (not just STDOUT)
		Ben April
0_1_3
	-Made customer name and number options.
	-Made name of "Customer Name" field an option.
	-Made name of "Customer Number" field an option.
	-Converted to DBI.
	-rev_gen_tool.pl now creates real config values for config.file.
	-Patch from James Raftery (HTML and moved some functions.)
	-Patch from Dave Burke (HTML fixes. Moved some HTML values to config.)
	-ipdb_ck.pl's main pass is now recursive.
	-if no block are found the Search function becomes recursive.
	-The Search function can now look for addresses that are not on bit-boundaries.
	-The Admin page is more verbose.
	-Fixed some bugs in search.cgi
0_1_4
	-Piles of bug-fixes.
	-Added report.pl
	-Display DNS in search.
	-Add option to toggle DNS zones on and off.
	-Reclaim stops at region boundies.
	-Change "= NULL" to "IS NULL" [ Martin Richard ]
	-Patch from Stojan Rancic
	-Dave Burke \H fix.
	-Montgomery Newcom DBI::err fix
	-Added import tool.
	-Hostnames are now searchable.

0_1_5
	-Function to delete RA
	-Function to delete Region
	-Test for Math::BigInt


test.pl
#!/usr/bin/perl -w

use strict;
use Math::BigInt;

# Just for fun, here's an example bit of code.  It should print "1\n" and
# exit almost immediately.  If this code hangs, you need to upgrade your
# Math::BigInt:
# -monte

print "\n\n --- Math::BigInt (blsft) test ---\n\n";
print "If you don't see \"Done\"";
print " in a few seconds try a diffrent version of Math::BigInt\n";

my $temp = Math::BigInt->new(1);
$temp = $temp->blsft(0);
print "$temp\n";

print "Done. You're good to go!\n";
print "\n";



ipdb_ck.pl
#!/usr/bin/perl
#                _____              ___ ____     _ _
#               |  ___| __ ___  ___|_ _|  _ \ __| | |__
#               | |_ | '__/ _ \/ _ \| || |_) / _` | '_ \
#               |  _|| | |  __/  __/| ||  __/ (_| | |_) |
#               |_|  |_|  \___|\___|___|_|   \__,_|_.__/
#
#	ipdb_ck.pl-$Name:  $-$Revision: 1.10 $ $Date: 2002/03/10 03:53:58 $ <$Author: bapril $@freeipdb.org>
######################################################################


use config;
use DBI;
%config = config::config();
require 'ipdb_lib.pl';

my $conn = DBI->connect("DBI:Pg:dbname=$config{dbname};host=$config{dbhost};port=$config{dbport}", "$config{dbuser}", "$config{dbpass}",);

print " /*--------------------- IP Database Consistency Check ---------------------*/\n\n";


$change = 1;
while($change){
	#Check for reclaims that haven't rolled up due to holdtime.
	$now = time;
	$query = "SELECT ID,PARENT,REGION,BITS
		FROM IPDB
		WHERE ( HOLDTIME < $now OR HOLDTIME IS NULL)
			AND ALLOCATED IS NULL 
			AND CHILDL IS NULL
			AND (RECLAIM != 1 OR RECLAIM IS NULL)";
	my $str = $conn->prepare($query);
	my $num = $str->execute;
	$first = 0;
	$change = 0;
	$i = 0;
	my @out;
	while(@out = $str->fetchrow){
		$id = $out[0];
		$parent = $out[1];
		$region = $out[2];
		$bits = $out[3];
		unless($PRTNS[$parent]){
			$FIRST[$parent] = $id;
		}
		$PRTNS[$parent]++;
		if($PRTNS[$parent] == 2){
			unless($first){print "reclaimed pairs that need to be cleared.\n";$first = 1;}
			print "ID $id - $FIRST[$parent] Cleared. $parent\n";
			demoblock($conn,$id) || &IPDBError(-1,"Could not demoblock");
			demoblock($conn,$FIRST[$parent]) || &IPDBError(-1,"Could not demoblock #2");
			clearblock($conn,$parent,1) || &IPDBError(-1,"Could not clearblock #3");
			$change++;
		}
		$i++;
	}
}

# List blocks that are still in holdtime.
$now = time;
$str = $conn->prepare("SELECT BLOCK,BITS,HOLDTIME
	FROM IPDB
	WHERE HOLDTIME > $now 
	AND ALLOCATED IS NULL
	AND CHILDL IS NULL ORDER BY HOLDTIME");
$num = $str->execute;
$first = 0;
$i = 0;
my @out;
while(@out = $str->fetchrow){
	unless($first){print "reclaimed blocks in holdtime.\n";$first = 1;}
	$block = $out[0];
	$bits = $out[1];
	$holdtime = $out[2];
	$diff = $holdtime - $now;
	if($diff > 86400 && $olddiff < 86400){print "---- One Day ----\n";}
	if($diff > 172800 && $olddiff < 172800){print "---- Two Days ----\n";}
	if($diff > 259200 && $olddiff < 259200){print "---- Three Days ----\n";}
	if($diff > 345600 && $olddiff < 345600){print "---- Four Days ----\n";}
	if($diff > 432000 && $olddiff < 432000){print "---- Five Days ----\n";}
	if($diff > 518400 && $olddiff < 518400){print "---- Six Days ----\n";}
	if($diff > 604800 && $olddiff < 604800){print "---- Seven Days ----\n";}
	if($diff > 691200 && $olddiff < 691200){print "---- Eight Days ----\n";}
	if($diff > 775800 && $olddiff < 775800){print "---- Nine Days ----\n";}
	if($diff > 864000 && $olddiff < 864000){print "---- Ten Days ----\n";}
	$olddiff = $diff;
	$block = deci2ip($block,4);
	printf("%20s Time: %s\n ",$block."/".$bits,$diff);
	$i++;
}

# Check for parents with only one child defined.
$str = $conn->prepare("SELECT ID FROM IPDB WHERE (CHILDL IS NULL AND CHILDR NOTNULL) OR (CHILDR IS NULL AND CHILDL NOTNULL)");
$num = $str->execute;
$i = 0;
if($num){print "Blocks that have only one child:\n";}
while(@out = $str->fetchrow){
	$id = $out[0];
	print "\t$id\n";
	$i++;
}

#Starting with the smallest block(s)
	#Gen list of block and thei parrents.
	#walk the list
		#Query parent
			#confirm children are truly related.
			#confirm any siblings.
			#Remove any ok siblings
			#Add parents to the list of bits--
		#
	#
#Incrament blocksize.

$str->finish;
undef $str;
$conn->disconnect;
print "Done\n";



rev_gen_tool.pl
#!/usr/local/bin/perl
#                _____              ___ ____     _ _
#               |  ___| __ ___  ___|_ _|  _ \ __| | |__
#               | |_ | '__/ _ \/ _ \| || |_) / _` | '_ \
#               |  _|| | |  __/  __/| ||  __/ (_| | |_) |
#               |_|  |_|  \___|\___|___|_|   \__,_|_.__/
#
#	rev_gen_tool.pl-$Name:  $-$Revision: 1.14 $ $Date: 2002/04/29 19:15:38 $ <$Author: bapril $@freeipdb.org>
######################################################################

# Trial tool to generate Reverse DNS files from IPDB/DNS.
require 'ipdb_httpcgi.pl';
use config;
%config = config::config();

$SIG{__WARN__} = sub {my$d=join('',@_);my$l=$d=~s/ at (.*) line (\d+)(?:, <\w+> chunk \d+)?\.$//?"$1($2): ":'';print"$l$_\n"for split(/\n/,$d)}; 

my $debug = 1;
$force = 0;
$quiet = 0;
foreach $arg (@ARGV){
	if($arg eq "-q"){ $quiet = 1;}
	if($arg eq "-f"){ $force = 1;}
	if($arg eq "-d"){ $debug = 1;} else { $debug = 0;}
	if($arg eq "-?" || $arg eq "-h"){
		print "Usage: rev_gen_tool.pl [-q] [-h]\n";
		print "\t-h this help screen.\n";
		print "\t-q operate queitly (for cron and such).\n";
		exit();
	}
}

unless($config{allowDNS}){
	die("FreeIPdb is not configured for DNS.\n")
}

#Check for DNS dir
open(CONFIG,">$config{dnsdir}/$config{dnsinclude}")|| die "Can't open $config{dnsdir}/$config{dnsinclude} for writing $!";

use DBI;
my $conn = DBI->connect("DBI:Pg:dbname=$config{dbname};host=$config{dbhost};port=$config{dbport}", "$config{dbuser}", "$config{dbpass}",);

#Walk DB for Reverse Files.

my $query = "SELECT ZONE_TABLE.ID,ZONE_TABLE.reverse_block WHERE ZONE_TABLE.reverse_block NOTNULL AND IPDB.ID = ZONE_TABLE.reverse_block AND IPDB.ID NOTNULL AND ZONE_TABLE.OWN = 't' ORDER BY IPDB.block";
my $str = $conn->prepare($query);
my $num = $str->execute;
my @out;
while(@out = $str->fetchrow){
	my $id = $out[0];
	my $rev_zone = $out[1];
	my $zonename = &GetZoneString($conn,$id);
	$zonename =~ s/.$//g;
	my $fileserial = &GetFileSerial("$config{dnsdir}$zonename");
	my $zoneserial = &GetZoneSerial($conn,$rev_zone,1);
	warn "ZONE: $zonename($rev_zone)  Serial file/zone: $fileserial/$zoneserial" if $debug;
	if($fileserial != $zoneserial || $force){
		open(FILE,">$config{dnsdir}$zonename")|| die "Can't open $config{dnsdir}$zonename for writing $!";
		select(FILE);
		&DNSHead($conn,$rev_zone,1);
		# Walk blocks under zone.
		&WalkReverseBlocks($conn,$rev_zone);
		select(STDOUT);
		close(FILE);
		warn "ZONE $zonename" unless $quiet;
		print CONFIG "zone \"$zonename\" in {\n";
		print CONFIG "\ttype master;\n";
		print CONFIG "\tfile \"$config{dnsdir}$zonename\";\n";
		print CONFIG "};\n";
	} else {
		warn "FILE: $fileserial | DB: $zoneserial" if $debug;
	}
}


#Walk DB for Forward File.

$query = "SELECT ID FROM ZONE_TABLE WHERE OWN = 't' AND reverse_block IS NULL";
$str = $conn->prepare($query);
$num = $str->execute;
while(@out = $str->fetchrow){
	my $id = $out[0];
	my $zonename = &GetZoneString($conn,$id);
	$fileserial = &GetFileSerial("$config{dnsdir}$zonename");
	$zoneserial = &GetZoneSerial($conn,$id,0);
	if($fileserial != $zoneserial || $force){
		open(FILE,">$config{dnsdir}$zonename")|| die "Can't open $config{dnsdir}$zonename for writing $!";
		select(FILE);
		# Walk the Zone.
		&DNSHead($conn,$id);
		&WalkZone($conn,$id);
		select(STDOUT);
		close(FILE);
		warn "ZONE $zonename" unless $quiet;
		print CONFIG "zone \"$zonename\" in {\n";
		print CONFIG "\ttype master;\n";
		print CONFIG "\tfile \"$config{dnsdir}$zonename\";\n";
		print CONFIG "};\n";
	}
}


close(CONFIG);
$str->finish;
undef $str;
$conn->disconnect;


report.pl
#!/usr/local/bin/perl -w
#                _____              ___ ____     _ _
#               |  ___| __ ___  ___|_ _|  _ \ __| | |__
#               | |_ | '__/ _ \/ _ \| || |_) / _` | '_ \
#               |  _|| | |  __/  __/| ||  __/ (_| | |_) |
#               |_|  |_|  \___|\___|___|_|   \__,_|_.__/
#
#       report.pl-$Revision: 1.3 $ $Date: 2002/05/07 23:30:56 $ <$Author: bapril $@freeipdb.org>
######################################################################

require 'ipdb_httpcgi.pl';

use config;
use DBI;
%config = config::config();

my $conn = DBI->connect("DBI:Pg:dbname=$config{dbname};host=$config{dbhost};port=$config{dbport}", "$config{dbuser}", "$config{dbpass}",);

print "\n/----------------------  IP Allocation Utilization  ----------------------/\n";
&UtilDisplay($conn);
print "\n\n";

print "\n/----------------------  IP Region Utilization  ----------------------/\n";
&UtilRegion($conn);
print "\n\n";


if($config{routeview}){
	use Net::Telnet();
	print "\n/----------------------  Reclaimed blocks being announced ----------------------/\n";
	&routeview_report($conn);
}

$conn->disconnect;
